### PR TITLE
Massive model update on the web scraping engine

### PR DESCRIPTION
**Pull request by jayzz55**
I have re-built and code everything from scratch on the web scraping engine. This time using the sidekiq integrated with Rails active jobs, and also a much better algorithm that check the catalogues before deciding to start scraping the site. These are the main features on this massive model update:
- Introduce Catalogue model
- Integrating Sidekiq together with Active jobs for a queue job with 1 worker working in the background
- an algorithm that check existing catalogue data against the data live on the site. this will delete old catalogue data, duplicate data, or add new catalogue data in the database.

### CODE REVIEW

File: app/jobs/catalogue_check_job.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
+class CatalogueCheckWorker < ActiveJob::Base
+  queue_as :default
+
+  require 'capybara'
+  require 'capybara/poltergeist'
+  require_relative '../../app/jobs/scrape_job.rb'
+  require_relative '../../app/jobs/delete_job.rb'
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:7]: Does this work without the requires for ScrapeJob and DeleteJob?

File: app/jobs/catalogue_check_job.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
+class CatalogueCheckWorker < ActiveJob::Base
+  queue_as :default
+
+  require 'capybara'
+  require 'capybara/poltergeist'
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:5]: In the `Gemfile`, move the capybara and poltergeist gems so they aren't within any group, as we want them available in all environments.

Does this work without the requires on lines 4 and 5?

File: app/jobs/catalogue_check_job.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  include Capybara::DSL
+  Capybara.register_driver :poltergeist do |app|
+    Capybara::Poltergeist::Driver.new(app, timeout: 60, js_errors: false)
+  end
+  Capybara.default_driver = :poltergeist
+
+  def perform
+
+    #setting the initial variable
+    current_catalogue = Hash.new(0)
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:18]: Does the job make use of the default `0` value setup here in the `Hash` initializer?

File: app/jobs/catalogue_check_job.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    compare_catalogue_array = []
+    new_catalogue = []
+    catalogue_delete = []
+    region_array = ["Melbourne, 3000", "Sydney, 2000"]
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:22]: `regions` would be a more concise name than `region_array`. Avoid putting data type in the variable name most of the time.

File: app/jobs/catalogue_check_job.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    Capybara::Poltergeist::Driver.new(app, timeout: 60, js_errors: false)
+  end
+  Capybara.default_driver = :poltergeist
+
+  def perform
+
+    #setting the initial variable
+    current_catalogue = Hash.new(0)
+    compare_catalogue_array = []
+    new_catalogue = []
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:20]: `new_catalogues` would be a slightly more accurate name.

File: app/jobs/catalogue_check_job.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        page.find('input#location-search').set(region)
+        page.find('div.autocomplete-suggestion strong').click
+
+        #scrape the catalogue number information
+        item_url = page.find('a.catalogue-image img')[:src]
+        start_count = "http://salefinder.com.au/images/featureimage/iphone/".length
+        end_count = item_url.length - (".jpg".length) - 1
+        array_result <<  catalogue_num = item_url.slice!(start_count..end_count)
+
+        sleep 5
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:71]: Avoid magic numbers. Assign `5` to a variable to describe what it is.

File: app/jobs/catalogue_check_job.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  def perform
+
+    #setting the initial variable
+    current_catalogue = Hash.new(0)
+    compare_catalogue_array = []
+    new_catalogue = []
+    catalogue_delete = []
+    region_array = ["Melbourne, 3000", "Sydney, 2000"]
+
+    #collect current_catalogue data
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:24]: The readability of the catalogue processing code here could be improved. It's a little hard to follow exactly what's happening. Once there are some tests in place for this code, it'd be a good candidate for refactoring. 

I like what you've done with extracting the `catalogue_gather` method. You can extract more methods here and give them good names, that'd help readability and avoid the need for the comments. Aim to have well named methods and variables to avoid the need for explanatory comments.

File: app/jobs/catalogue_check_job.rb (hunk 4 above)
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:22]: Remove region_array from this method and move it into the `catalogue_gather` method. The `perform` method doesn't need to know about it.

File: app/jobs/catalogue_check_job.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    #delete old and duplicate catalogues
+    catalogue_delete = old_data + duplicate_data
+    DeleteWorker.perform_later(catalogue_delete)
+  end
+
+  private
+
+  def catalogue_gather(region_array)
+    array_result = []
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:53]: Remove region_array argument from this method and just declare it inside the method. `region_array` only needs to be known in this method, avoid exposing it elsewhere unnecessarily.

Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:53]: Come up with a more descriptive name than `array_result`

File: app/jobs/catalogue_check_job.rb [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    region_array.each do |region|
+      ['Woolworths', 'Coles'].each do |shop|
+
+        #visit the site
+        visit "http://salefinder.com.au/#{shop}-catalogue?qs=1,,0,0,0"
+
+        #set the region of the catalogue we want to gather
+        page.find('a#header-change-region').trigger('click')
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:61]: Does `.click` work here instead of `.trigger('click')`?

File: app/jobs/catalogue_check_job.rb [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        page.find('input#location-search').set(region)
+        page.find('div.autocomplete-suggestion strong').click
+
+        #scrape the catalogue number information
```
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:65]: These lines are another candidate for extracting into a separate method. Giving that method a good name will mean the comment on line 65 can be removed.

File: app/jobs/delete_job.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class DeleteWorker < ActiveJob::Base
+  queue_as :default
+
+  include Sidekiq::Worker
+
+  def perform(catalogue_delete)
+    catalogue_delete.each{|c| Catalogue.find_by(catalogue_num:c).destroy}
```
Comment (eliotsykes) [inline app/jobs/delete_job.rb:7]: Consider renaming `catalogue_delete` to `catalogue_nums` and change line 7 to the more succinct:

``` ruby
Catalogue.where("catalogue_num in ?", catalogue_nums).destroy_all
```

File: app/jobs/delete_job.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+class DeleteWorker < ActiveJob::Base
```
Comment (eliotsykes) [inline app/jobs/delete_job.rb:1]: A name like `DeleteCataloguesJob` would reveal the intention of this class more clearly.

File: app/jobs/clean_job.rb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+class CleanupWorker < ActiveJob::Base
```
Comment (eliotsykes) [inline app/jobs/clean_job.rb:1]: Try to have the class name and file name match. Consider renaming this to `DeleteAllCataloguesJob` to accurately describe what it does.

File: app/jobs/clean_job.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+class CleanupWorker < ActiveJob::Base
+  queue_as :default
+
+  include Sidekiq::Worker
```
Comment (eliotsykes) [inline app/jobs/clean_job.rb:4]: Is `include Sidekiq::Worker` still needed in the 3 jobs its declared in?

File: app/jobs/scrape_job.rb [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+  require 'capybara'
+  require 'capybara/poltergeist'
+
+  include Sidekiq::Worker
+  include Capybara::DSL
+  Capybara.register_driver :poltergeist do |app|
+    Capybara::Poltergeist::Driver.new(app, timeout: 60, js_errors: false)
+  end
+  Capybara.default_driver = :poltergeist
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:12]: The capybara config here on lines 9 to 12 (**not** the include on line 8) might be better in an initializer in `config/initializers`. Consider moving this config out of the 2 jobs it is currently in and put it in a single initializer.

File: app/jobs/scrape_job.rb [hunk 16]
```diff
@@ -0,0 +1,66 @@
+class PostScrapeWorker < ActiveJob::Base
+  queue_as :default
+
+  require 'capybara'
+  require 'capybara/poltergeist'
+
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:6]: requires on line 4 and 5 needed?

File: app/jobs/scrape_job.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
+class PostScrapeWorker < ActiveJob::Base
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:1]: Have the file name and class name be in sync. Is there a name that reveals the purpose of this class a little clearer?

File: app/jobs/scrape_job.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  require 'capybara/poltergeist'
+
+  include Sidekiq::Worker
+  include Capybara::DSL
+  Capybara.register_driver :poltergeist do |app|
+    Capybara::Poltergeist::Driver.new(app, timeout: 60, js_errors: false)
+  end
+  Capybara.default_driver = :poltergeist
+
+  def perform(new_catalogue)
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:14]: `catalogue_nums` seems a clearer name that `new_catalogue`.

File: app/jobs/scrape_job.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    new_catalogue.each do |catalogue_num|
+      begin
+
+        #visit page
+        visit "http://salefinder.com.au/woolworths-catalogue/weekly-specials-catalogue/#{catalogue_num}/catalogue"
+        
+        #creating new catalogue
+        @catalogue = Catalogue.new
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:22]: No need for `@catalogue` to be an instance variable. Try to minimize variable scope where possible.

File: app/jobs/scrape_job.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        @catalogue.title = page.all('div#breadcrumb li a')[2].text
+        @catalogue.date = page.find('span.sale-dates').text
+        @catalogue.catalogue_num = catalogue_num
+        @catalogue.shop = page.all('div#breadcrumb li a')[1].text
+        @catalogue.shop.slice!(" catalogues")
+        if @catalogue.shop == "Woolworths"
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:28]: What do you think about introducing a `Shop` model to represent the shop specific data?

File: app/jobs/scrape_job.rb [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+          @catalogue.shop_logo = "http://salefinder.com.au/images/retailerlogos/126.jpg"
+        else
+          @catalogue.shop_logo = "http://salefinder.com.au/images/retailerlogos/148.jpg"
+        end
+        @catalogue.save
+
+        #start scraping from page 1 and loop to last page
+        page_num = 1
+        visit "http://salefinder.com.au/woolworths-catalogue/weekly-specials-catalogue/#{catalogue_num}/list?qs=#{page_num}"
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:37]: Extract the shared part of the URL into a variable to be used on line 19 and 37, 59. Perhaps introduce method(s) to build URL(s).

File: app/jobs/scrape_job.rb [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        while page.has_css?('span.price')
+          #get the price information data on each item on a page
+          all('div.item-landscape').each do |item|
+            description = item.find('span.item-details h1 a').text
+            price_info = item.first('span.price').text if item.first('span.price')
+            unit_price = item.first('span.comparative-text').text if item.first('span.comparative-text')
+            saving_info = item.find('div.price-options').text if item.find('div.price-options')
+            saving_info.slice!(price_info) if price_info
+            saving_info.slice!(unit_price) if unit_price
+            saving_info if item.find('div.price-options')
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:47]: What's line 47 doing?

File: app/jobs/scrape_job.rb [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+            Catalogue.find_by(catalogue_num: catalogue_num ).posts.create(description: description, price_info: price_info, unit_price: unit_price, saving_info: saving_info, image: image)
+          end
+          page_num += 1
+
+          #Capybara reset session
+          Capybara.reset_sessions!
+          page.driver.reset!
+
+          #visit the next page
+          sleep 5
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:58]: Magic number

File: app/jobs/scrape_job.rb [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+          visit "http://salefinder.com.au/woolworths-catalogue/weekly-specials-catalogue/#{catalogue_num}/list?qs=#{page_num}"
+        end
+      rescue Capybara::Poltergeist::TimeoutError 
+        retry
```
Comment (eliotsykes) [inline app/jobs/scrape_job.rb:62]: The job could run indefinitely. Limit the number of retries.

File: exit [hunk 25]
```diff
@@ -0,0 +1,88 @@
+=> [[32m#<Catalogue:0xbd9fde0[0m
```
Comment (eliotsykes) [inline exit:1]: This file probably isn't needed in the repo.

File: lib/tasks/scraper.rake [hunk 26]
```diff
@@ -1,101 +1,18 @@
...
-
-    rescue Capybara::Poltergeist::TimeoutError 
-      p "!!!!!!!!!!!!!!!------rescue is working on page #{page_number}-----!!!!!!!!!!!!!!!!!!!!!!"
-      num_rescue = num_rescue + 1
-      retry
-
-    end
+  desc "Destroy all posting data"
+  task destroy_data: :environment do
+    require_relative '../../app/jobs/clean_job.rb'
```
Comment (eliotsykes) [inline lib/tasks/scraper.rake:5]: Is the require needed here?

File: app/jobs/catalogue_check_job.rb (hunk 3 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:18]: I used Hash.new(0), because I came across this post: http://stackoverflow.com/questions/2698460/strange-ruby-behavior-when-using-hash-new
So i thought that it'll be safer to use this to avoid unwanted behavior.

any thought on better way to do this?

File: app/jobs/catalogue_check_job.rb (hunk 1 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:7]: I've tested this, without the require for ScrapeJob and DeleteJob, the program will not work as it doesn't recognize the class.

File: lib/tasks/scraper.rake (hunk 26 above)
Comment (jayzz55) [inline lib/tasks/scraper.rake:5]: Yes, without the require, the program will not recognize the class defined in clean_job.rb

File: app/jobs/scrape_job.rb (hunk 22 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:47]: This has been fixed in #4 Pull Request. line 47 code is a redundant code.

File: app/jobs/scrape_job.rb (hunk 20 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:28]: I agree that having a `shop` model would model the ideal relationship in this data. When I built this, I'm looking for the minimum model to get it running and also in this program, we only have 2 shops. so at this point of time, I decided to stick without the `shop` model, and implement it later in the future.

File: app/jobs/clean_job.rb (hunk 14 above)
Comment (jayzz55) [inline app/jobs/clean_job.rb:4]: fixed in commit `remove include Sidekiq::Worker` in PR#5 Fix-comment-1

File: app/jobs/catalogue_check_job.rb (hunk 2 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:5]: I've made capybara and poltergeist gem available in all environment in PR #5. However, i found that the code is not working without requires on line 4 and line 5.

File: app/jobs/catalogue_check_job.rb (hunk 4 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:22]: Fixed this in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 5 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:20]: Fixed in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 9 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:61]: Unfortunately no, because the way the site overlay the object to be clicked. When I run `.click`, Capybara recommend to use `.trigger('click')` to trigger the javascript to show the object. without `.trigger('click')`, the element will be hidden, and Capybara will output error in finding the element.

File: app/jobs/catalogue_check_job.rb (hunk 6 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:71]: Fixed this in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 7 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:24]: Fixed this in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 8 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:53]: Fixed this in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 10 above)
Comment (jayzz55) [inline app/jobs/catalogue_check_job.rb:65]: The reason why I didn't extract this any further is to avoid extracting a capybara flow session into a separate method. since it's a one continuous flow, I think it's best to keep it as it is. any thought on this?

File: app/jobs/delete_job.rb (hunk 11 above)
Comment (jayzz55) [inline app/jobs/delete_job.rb:7]: I tried this and run into the following error:
`ActiveRecord::StatementInvalid: SQLite3::SQLException: near "NULL": syntax error: SELECT "catalogues".* FROM "catalogues" WHERE (catalogue_num in NULL)`

will need help to revise this.

File: app/jobs/delete_job.rb (hunk 12 above)
Comment (jayzz55) [inline app/jobs/delete_job.rb:1]: Fixed this in PR #5

File: app/jobs/clean_job.rb (hunk 13 above)
Comment (jayzz55) [inline app/jobs/clean_job.rb:1]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 15 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:12]: I'm not sure how to do this. will need help to play around with the `config/initializers`

File: app/jobs/scrape_job.rb (hunk 16 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:6]: yes, I can't get it running without requiring these

File: app/jobs/scrape_job.rb (hunk 17 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:1]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 18 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:14]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 19 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:22]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 21 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:37]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 23 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:58]: Fixed this in PR #5

File: app/jobs/scrape_job.rb (hunk 24 above)
Comment (jayzz55) [inline app/jobs/scrape_job.rb:62]: Fixed this in PR #5

File: exit (hunk 25 above)
Comment (jayzz55) [inline exit:1]: Fixed this in PR #5

File: app/jobs/catalogue_check_job.rb (hunk 3 above)
Comment (eliotsykes) [inline app/jobs/catalogue_check_job.rb:18]: The conventional way to initialize an empty `Hash` is to use the hash literal `{}`:

``` ruby
# Prefer {} to Hash.new
current_catalogue = {}
```

I think that'll work unless there's a good reason for the current_catalogue to have a default of `0`. Consider reading http://ruby-doc.org/core-2.2.0/Hash.html#method-c-new for more info on initializing Hashes with default values and blocks.
